src/net/socket/udp_sock.c

```diff
@@ -28,79 +29,133 @@
 #include <stdlib.h>
 
 static const struct sock_proto_ops udp_sock_ops_struct;
-const struct sock_proto_ops *const udp_sock_ops
-		= &udp_sock_ops_struct;
+const struct sock_proto_ops *const udp_sock_ops = &udp_sock_ops_struct;
 
 EMBOX_NET_SOCK(AF_INET, SOCK_DGRAM, IPPROTO_UDP, 1,
 		udp_sock_ops_struct);
 
-static int udp_sendmsg(struct sock *sk, struct msghdr *msg, int flags) {
-	size_t total_len, data_len, actual_len, skb_len;
-	struct sk_buff *skb;
-	const struct sockaddr_in *to;
-	const struct sockaddr *sockaddr;
-	int i, ret;
-
-	assert(sk);
-	assert(sk->o_ops);
-	assert(sk->o_ops->make_pack);
-	assert(msg);
-	assert(msg->msg_iov);
-	assert(msg->msg_iov->iov_base);
-
-	data_len = 0;
-	for (i = 0; i < msg->msg_iovlen; i++) {
-		data_len += msg->msg_iov[i].iov_len;
-	}
+static int iov_msg_len(struct iovec *iov, int iovlen) {
+	int len = 0;
 
-	total_len = data_len + UDP_HEADER_SIZE;
+	for (int i = 0; i < iovlen; i++) {
+		len += iov[i].iov_len;
+	}
 
-	skb = NULL;
-	sockaddr = (const struct sockaddr *)msg->msg_name;
+	return len;
+}
 
-	if (msg->msg_name != NULL) {
-		to = (const struct sockaddr_in *)msg->msg_name;
-	} else {
-		to = (const struct sockaddr_in *)&to_inet_sock(sk)->dst_in;
-	}
+int iov_stream_make_queue(size_t len, size_t hdr_size, struct sk_buff_head *queue,
```

Comment:
allocate and fill separation just makes sure you will not wast efforts on filling packets. Suppose you're doing allocate+fill steps as a part of big transmission: allocate+fill, allocate+fill, allocate+fill, ... until some allocation fails. Then you have to throw away all packets, because you're not going to send any of them. It's not matters, have you filled them or not.

More dangerous is to allocate+fill+send step as a part of transmission. If some allocation fail, then you shouldn't sent any packet that you've sent already, you couldn't revert them.
And this situation took place pefore the patch, on line 68 there was a loop on msg_iovlen, where `make_pack` and then `snd_pack` called in same iteration.